### PR TITLE
HIVE-20917: Support applyQuotesToAll option in OpenCSVSerde.

### PR DESCRIPTION
**Pull request by gigem**
OpenCSV version 3.10 added an option called "applyQuotesToAll."  When set to false, output columns were not quoted unless they needed it. This commit adds support for that option in OpenCSVSerde and also updates the support for OpenCSV version 5.7.1, which is the current version at the time of this commit.  This commit is based on a patch from Sungwoo Park <glapark@gmail.com>, which was based on a patch from David Engel <david@istwok.net>.

### CODE REVIEW

**Comment by BsoBird** [comment]
@deniskuzZ @ayushtkn  Hello. can we reopen this?

File: serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java [hunk 1]
```diff
@@ -90,6 +95,8 @@ public void initialize(Configuration configuration, Properties tableProperties,
     separatorChar = getProperty(properties, SEPARATORCHAR, CSVWriter.DEFAULT_SEPARATOR);
     quoteChar = getProperty(properties, QUOTECHAR, CSVWriter.DEFAULT_QUOTE_CHARACTER);
     escapeChar = getProperty(properties, ESCAPECHAR, CSVWriter.DEFAULT_ESCAPE_CHARACTER);
+    String temp = properties.getProperty(APPLYQUOTESTOALL);
+    applyQuotesToAll = (temp == null || Boolean.parseBoolean(temp));
```
Comment (deniskuzZ) [inline serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java:99]: you could just inline 
````
applyQuotesToAll = Boolean.parseBoolean(properties.getProperty(APPLYQUOTESTOALL))
````

File: serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java [hunk 2]
```diff
@@ -178,19 +185,26 @@ public Object deserialize(final Writable blob) throws SerDeException {
...
+    CSVParser parser;
     if (CSVWriter.DEFAULT_ESCAPE_CHARACTER == escape) {
-      return new CSVReader(reader, separator, quote);
+      parser = new CSVParserBuilder()
+          .withSeparator(separator)
+          .withQuoteChar(quote)
+          .build();
     } else {
-      return new CSVReader(reader, separator, quote, escape);
+      parser = new CSVParserBuilder()
```
Comment (deniskuzZ) [inline serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java:195]: can we avoid duplication by creating a builder var or using ternary operator? the only diff here is the missing escape char setup

File: serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java [hunk 3]
```diff
@@ -178,19 +185,26 @@ public Object deserialize(final Writable blob) throws SerDeException {
...
+        .build();
   }
 
   private CSVWriter newWriter(final Writer writer, char separator, char quote, char escape) {
-    if (CSVWriter.DEFAULT_ESCAPE_CHARACTER == escape) {
-      return new CSVWriter(writer, separator, quote, "");
-    } else {
-      return new CSVWriter(writer, separator, quote, escape, "");
-    }
+    return new CSVWriter(writer, separator, quote, escape, "");
```
Comment (deniskuzZ) [inline serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java:207]: why do we change the escape char handling here?

**Pull request closed without merging**